### PR TITLE
Added Job submission support to the API server

### PR DESCRIPTION
**Pull request by blublinsky**
## Why are these changes needed?

API server already provides support for Ray Job, but its usage is not very convenient for Python developer.  Its not easy to get execution log, stop running job, etc. This functionality is provided by Ray job submission API, but requires knowledge about endpoint and corresponding security when used outside the cluster. This implementation provides all the functionality through the API server (with the well known URL) and corresponding Python client

## Checks

- [x ] I've made sure the tests are passing. 
- Testing Strategy
   - [x ] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by tedhtchang — COMMENTED** [review]
Looks good. I was able to follow the README to test the PR. There are some typos.

File: apiserver/JobSubmission.md
Comment (tedhtchang) [inline apiserver/JobSubmission.md:18]: This instruction is good for development such as testing the PR.
```suggestion
For development, we can build the API server from source code and deploy to the local KinD cluster.
```shell
make docker-image cluster load-image deploy
```

```

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:3]: Fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:85]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:13]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:109]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:195]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:160]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:211]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:7]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:18]: fixed

**Comment by blublinsky** [comment]
Thank you @tedhtchang. Can you, please, approve it

**Review by tedhtchang — APPROVED** [review]

File: apiserver/JobSubmission.md
Comment (astefanutti) [inline apiserver/JobSubmission.md:13]: KubRay -> KubeRay

File: apiserver/JobSubmission.md
Comment (astefanutti) [inline apiserver/JobSubmission.md:147]: nit: missing trailing column.

File: apiserver/JobSubmission.md
Comment (astefanutti) [inline apiserver/JobSubmission.md:120]: nit: missing trailing column.

File: apiserver/JobSubmission.md
Comment (astefanutti) [inline apiserver/JobSubmission.md:160]: nit: missing trailing dot.

**Review by astefanutti — APPROVED** [review]

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:13]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:147]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:120]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:160]: fixed

**Comment by blublinsky** [comment]
@kevin85421, I think its ready for commit

**Comment by kevin85421** [comment]
@architkulkarni can review this PR. I will review it next week due to being overloaded this week.

File: clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py
Comment (architkulkarni) [inline clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py:20]: Nit: Why use -1 here? In the Ray Job API, if these fields are not specified, it defaults to 0.

File: clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py
Comment (architkulkarni) [inline clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py:88]:
```suggestion
        self.start_time = int(dst.get("startTime", "0"))
        self.end_time = int(dst.get("endTime", "0"))
```

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (architkulkarni) [inline ray-operator/controllers/ray/utils/dashboard_httpclient.go:472]: Question: Here and in general, is it better to raise an error if the job is not found?

File: clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py
Comment (architkulkarni) [inline clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py:108]: Can we add unit tests for `RayJobRequest.to_string`, `RayJobRequest.to_dict` and `RayJobInfo.to_string`?

**Review by architkulkarni — COMMENTED** [review]
Looks good overall, just a few minor questions.

@blublinsky From what I can see it seems like all of Ted's doc comments haven't been fixed actually (maybe forgot to push or something got messed up when merging?)

**Review by kevin85421 — COMMENTED** [review]
There are some issues that we need to figure out before merging this PR. @architkulkarni is the expert on both Ray and KubeRay. @architkulkarni, could you collaborate with @blublinsky to solve these? Thanks!

(1) Some fields are required in `RayJobInfo` and `RayJobRequest` in Ray's definition. However, all fields in the struct are optional.
(2) Double-check the types between KubeRay's `dashboard_httpclient.go` and Ray. For example, the type of `start_time` is `uint64` in Ray, but it is `int64` in this PR, which may cause data loss.
(3) Some fields are missing in the dashboard client.
(4) Is there any change in compatibility between Ray and KubeRay?

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:3]: Done

File: apiserver/pkg/server/ray_job_submission_service_server.go
Comment (blublinsky) [reply apiserver/pkg/server/ray_job_submission_service_server.go:193]: Otherwise go compile complains

File: clients/python-apiserver-client/python_apiserver_client/kuberay_apis.py
Comment (blublinsky) [reply clients/python-apiserver-client/python_apiserver_client/kuberay_apis.py:271]: Its always return 200 on success. there is no 204 here, so it works

File: clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py
Comment (blublinsky) [reply clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py:20]: I did not want to print/include in dict things that are not defined

File: clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py
Comment (blublinsky) [reply clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py:88]: fixed

File: clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py
Comment (blublinsky) [reply clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py:108]: All of them  are part of test_submission

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:109]: fixed

File: apiserver/JobSubmission.md
Comment (blublinsky) [reply apiserver/JobSubmission.md:89]: fixed

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (blublinsky) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:472]: Fixed

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (blublinsky) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:335]: No, check here https://docs.ray.io/en/latest/cluster/running-applications/job-submission/api.html#/paths/~1api~1jobs/post

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (architkulkarni) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:319]: The GCS proto is an internal field so I don't think we need to consider it. For this PR, I think it's okay to just add a subset of commonly used fields in https://docs.ray.io/en/latest/cluster/running-applications/job-submission/doc/ray.job_submission.JobInfo.html#ray.job_submission.JobInfo . It's very easy to add more fields later if needed.

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (blublinsky) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:326]: fixed

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (architkulkarni) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:335]: Thanks for catching this, both of you.  `entrypoint_memory` is a new field in Ray 2.8, and it looks like we've forgotten to add it to the Ray Job OpenAPI docs (those docs are unfortunately generated manually). 

If we add it to this PR, the apiserver will be incompatible with older versions of Ray unless we add some logic to check the Ray version somehow (maybe with the `/version` endpoint).  I think we can leave this for a future PR.

Comment (architkulkarni) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:335]: Created an issue for the Ray doc: https://github.com/ray-project/ray/issues/41277

File: apiserver/JobSubmission.md
Comment (architkulkarni) [reply apiserver/JobSubmission.md:13]: Seems not fixed still? Also KubRay -> KubeRay

File: apiserver/JobSubmission.md
Comment (architkulkarni) [reply apiserver/JobSubmission.md:3]: @blublinsky these doc changes don't show up as fixed on Github, can you double check on your end?

Comment (architkulkarni) [reply apiserver/JobSubmission.md:3]: @blublinsky These fixes aren't showing up as fixed on Github, can you double check on your end? Are they being overwritten somehow?

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (blublinsky) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:325]: fixed

**Comment by blublinsky** [comment]
@kevin85421, @architkulkarni  addressed your comments

**Comment by kevin85421** [comment]
`e2e / ray-operator` fails. @blublinsky could you take a look?

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (blublinsky) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:319]: It requires quite a few changes in the existing code

File: ray-operator/controllers/ray/utils/dashboard_httpclient.go
Comment (astefanutti) [inline ray-operator/controllers/ray/utils/dashboard_httpclient.go:370]: That change breaks the RayJob controller, and the Ray cluster dashboard check that's performed prior to submitting the job, and that's been added as part of #1512, with a work-around for #1381.

I think the easier option in the short term is to update the call from the RayJob controller to check the error (better be NotFound error).

@kevin85421 @architkulkarni what do you think, you guys probably have more context on what's the best strategy to handle this?

Comment (blublinsky) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go:370]: Thank you @astefanutti. For the time being, I rolled back the changes (preserving them as comments). Lets merge this one and move the issue of `the RayJob controller to check the error` to a separate PR

Comment (architkulkarni) [reply ray-operator/controllers/ray/utils/dashboard_httpclient.go]: Yes, rolling back sounds good so we can merge this PR.  There is another PR in flight which is dealing with this issue: https://github.com/ray-project/kuberay/pull/1623

**Review by architkulkarni — APPROVED** [review]
@blublinsky It looks like most of the doc feedback for `JobSubmission.md` has been commented as "fixed" but is not actually fixed in the file. I'm okay with merging this PR to unblock your other PRs, but please help address the doc feedback either in this PR or a future PR.

File: clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py
Comment (architkulkarni) [inline clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py:98]:
```suggestion
        if self.start_teme > 0:
```
@blublinsky We still need to change this one to `start_time` and all the other occurrences, otherwise I think it will error.  Rhe unit test should have caught this.

Comment (architkulkarni) [reply clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py:98]: I see we are calling `RayJobInfo.to_string` in the unit test, so I'm confused how it could be passing.  Is is possible that it's not running in CI?

Comment (blublinsky) [reply clients/python-apiserver-client/python_apiserver_client/params/jobsubmission.py]: fixed

**Comment by blublinsky** [comment]
@kevin85421, can we merge, please

**Comment by kevin85421** [comment]
@architkulkarni would you mind answering https://github.com/ray-project/kuberay/pull/1639#pullrequestreview-1740469017 before we merge the PR? Thanks!

File: proto/job_submission.proto
Comment (architkulkarni) [inline proto/job_submission.proto:166]: `status` is required.

File: proto/job_submission.proto
Comment (architkulkarni) [inline proto/job_submission.proto:160]: `entrypoint` is required.

**Review by architkulkarni — COMMENTED** [review]
> (1) Some fields are required in RayJobInfo and RayJobRequest in Ray's definition. However, all fields in the struct are optional.
(2) Double-check the types between KubeRay's dashboard_httpclient.go and Ray. For example, the type of start_time is uint64 in Ray, but it is int64 in this PR, which may cause data loss.
(3) Some fields are missing in the dashboard client.
(4) Is there any change in compatibility between Ray and KubeRay?

(1) I have double checked them and indicated the necessary changes in the below comments.
(2) This was fixed in https://github.com/ray-project/kuberay/pull/1639/commits/4c12da6a7a29aa28d0854cd01c665becbc8d5ee0 and then un-fixed in https://github.com/ray-project/kuberay/pull/1639/commits/5bb51ccfdea3b1301492beca4c3bab2f19ee8a01, it's not clear why.  @blublinsky can you check this? I verified that a millisecond unix timestamp cannot overflow `int64`, but we might as well be consistent and use `uint64`.
(3) @kevin85421 which fields did you notice were missing? Are you talking about things like `driver_node_id` and `driver_exit_code` in `JobInfo`? These were missing before this PR and it's easy to add them later if needed. I don't think it needs to be part of this PR, but let me know if I misinterpreted your comment.
(4) No, the changes to the dashboard API client are all for Ray Job API methods that have been supported since Ray 2.2.0 (DELETE was the most recent one).  KubeRay 1.1.0 only supports Ray 2.8.0 and later, so this is fine.

File: proto/job_submission.proto
Comment (blublinsky) [reply proto/job_submission.proto:166]: We defined as required only the data that user provides data, doing it for data coming from the server does not make sense to me

**Comment by blublinsky** [comment]
> I have double checked them and indicated the necessary changes in the below comments.
> (2) This was fixed in [4c12da6](https://github.com/ray-project/kuberay/commit/4c12da6a7a29aa28d0854cd01c665becbc8d5ee0) and then un-fixed in [5bb51cc](https://github.com/ray-project/kuberay/commit/5bb51ccfdea3b1301492beca4c3bab2f19ee8a01), it's not clear why. @blublinsky can you check this? I verified that a millisecond unix timestamp cannot overflow `int64`, but we might as well be consistent and use `uint64`

These fields were already defined as int64. I did not introduce them. We can change them, but it creates more changes outside this PR

File: proto/job_submission.proto
Comment (blublinsky) [reply proto/job_submission.proto:160]: @architkulkarni we only provide required annotation for the user-defined fields to catch validation errors. This message is coming from the server, so such annotation is pointless

**Comment by architkulkarni** [comment]
> We defined as required only the data that user provides data, doing it for data coming from the server does not make sense to me

Ok, that makes sense. Since we're not sending this data to the server there's no downside to leaving it as optional.

> These fields were already defined as int64. I did not introduce them. We can change them, but it creates more changes outside this PR

That's fine with me, we can update it in a separate PR.

@kevin85421 thoughts? Is it okay to merge?

**Comment by architkulkarni** [comment]
Discussed offline with @kevin85421 , it's okay to merge. The failing test  Compatibility Test - Nightly (pull_request) is flaky and unrelated (FAIL: test_detached_actor (__main__.RayFTTestCase.test_detached_actor))

**Pull request merged**
